Add GCS support for document loading

SurajKamal24 opened the pull request:
This PR adds support for loading documents from Google Cloud Storage (GCS) for the operations "Embedding add document to store" and "Embedding add folder to store," complementing the existing S3 and Azure Blob functionality. Developers can now seamlessly load documents to vector stores/databases by configuring their GCS bucket and credentials. 

The implementation follows a similar pattern to the existing S3 and Azure Blob document loaders.

**Implementation Details**

1. Leveraged the GoogleCloudStorageDocumentLoader API from the langchain4j library to integrate GCS functionality.
2. Embedding add document to store operation - Loads a single document from the GCS bucket to the chroma store using the context path pattern: gcs://{bucketName}/{objectName} or gcs://{bucketName}/{folderName}/{fileName}
3. Embedding add folder to store operation - Loads multiple documents from a specific folder in the GCS bucket using the context path pattern: gcs://{bucketName}/{folderName}. Documents are loaded based on the folder's prefix.
4. Authentication uses Google Cloud's GoogleCredentials library via a service account key JSON file. The absolute path for this file must be defined in the connector configuration. Please find the screenshot below

![gcs](https://github.com/user-attachments/assets/b7c274fb-54f8-4993-bdf4-9d530fa7c439)


Performed tests using a live GCS bucket to validate Document loading, Path prefix handling, Error scenarios (e.g., invalid credentials, non-existent bucket, non-existent object).

tbolis-at-mulesoft commented:
Please modify the PR to merge into develop branch.
Did you tested also the Document Operations? They support cloud storage as well.
Can you provide an example of how the json file is structured?
Thank you for your contribution.

tbolis-at-mulesoft commented on src/main/java/org/mule/extension/vectors/internal/storage/gcs/GoogleCloudStorage.java:
@SurajKamal24 let's throw a ModuleException with type STORAGE_SERVICES_FAILURE. Otherwise the user will see generic DOCUMENT_OPERATIONS_FAILURE or EMBEDDING_OPERATIONS_FAILURE. I'm referring to getLoader().

SurajKamal24 commented:
Thanks @tbolis-at-mulesoft for your feedback. Please find my comments below:

1. Modified the PR to merge into the develop branch
2. Tested both document parsing and chunk splitting operations. They work without any issues
3. Updated the exception handling to throw ModuleException with type STORAGE_SERVICES_FAILURE for GCS connectivity failures
4. Below is the screenshot of the service account key JSON file. You will get this when you generate a key for the service account user in the Google Cloud Console.

![keyfile](https://github.com/user-attachments/assets/7b7dbbf1-8caa-4735-8385-e48aec13a50c)

tbolis-at-mulesoft commented:
Thank you for the contribution @SurajKamal24 !

tbolis-at-mulesoft commented:
@amirkhan-ak-sf I just figured out that with gpc the size of the connector is to bit and make any app deployment to ch2 impossible... we are now close 400 MB while before 327MB. If we do not find a solution we need to temporary remove this.

amirkhan-ak-sf commented:
Yes we need to go back to 326MB otherwise we may end up blocking deployment
to ch2.

Kind regards,

<http://www.mulesoft.com/de/>
*{ *
"*name*"*: "Amir Khan", *
"role": "Client Solution Engineer",

"address": "Ernst-Nobs-Platz 1, CH-8004 Zürich",

"*phone*": "+41(0)79 296-4609",

"*email*"*: ***@***.***"

*}*



On Fri, 29 Nov 2024 at 00:18, Tommaso Bolis ***@***.***>
wrote:

> @amirkhan-ak-sf <https://github.com/amirkhan-ak-sf> I just figured out
> that with gpc the size of the connector is too bit and make any app
> deployment to ch2 impossible... we are now close 400 MB while before 327MB.
> If we do not find a solution we need to temporary remove this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MuleSoft-AI-Chain-Project/mule-vectors-connector/pull/65#issuecomment-2506845285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUWB2F7GXQFR5IEZ7HQMTGD2C6QEXAVCNFSM6AAAAABSUNU43GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMBWHA2DKMRYGU>
> .
> You are receiving this because you were mentioned.Message ID:
> <MuleSoft-AI-Chain-Project/mule-vectors-connector/pull/65/c2506845285@
> github.com>
>

amirkhan-ak-sf commented:
The only thing would be to somehow dynamically load the required
dependencies at design time in Studio.

Kind regards,

<http://www.mulesoft.com/de/>
*{ *
"*name*"*: "Amir Khan", *
"role": "Client Solution Engineer",

"address": "Ernst-Nobs-Platz 1, CH-8004 Zürich",

"*phone*": "+41(0)79 296-4609",

"*email*"*: ***@***.***"

*}*



On Fri, 29 Nov 2024 at 03:24, Amir Khan ***@***.***> wrote:

> Yes we need to go back to 326MB otherwise we may end up blocking
> deployment to ch2.
>
> Kind regards,
>
> <http://www.mulesoft.com/de/>
> *{ *
> "*name*"*: "Amir Khan", *
> "role": "Client Solution Engineer",
>
> "address": "Ernst-Nobs-Platz 1, CH-8004 Zürich",
>
> "*phone*": "+41(0)79 296-4609",
>
> "*email*"*: ***@***.***"
>
> *}*
>
>
>
> On Fri, 29 Nov 2024 at 00:18, Tommaso Bolis ***@***.***>
> wrote:
>
>> @amirkhan-ak-sf <https://github.com/amirkhan-ak-sf> I just figured out
>> that with gpc the size of the connector is too bit and make any app
>> deployment to ch2 impossible... we are now close 400 MB while before 327MB.
>> If we do not find a solution we need to temporary remove this.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/MuleSoft-AI-Chain-Project/mule-vectors-connector/pull/65#issuecomment-2506845285>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AUWB2F7GXQFR5IEZ7HQMTGD2C6QEXAVCNFSM6AAAAABSUNU43GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMBWHA2DKMRYGU>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> <MuleSoft-AI-Chain-Project/mule-vectors-connector/pull/65/c2506845285@
>> github.com>
>>
>

tbolis-at-mulesoft commented:
@SurajKamal24 @amirkhan-ak-sf Development parked in [feature/add-gcs-support](https://github.com/MuleSoft-AI-Chain-Project/mule-vectors-connector/tree/feature/add-gcs-support). Let's see how we can dynamically add required dependencies.
